test/tap/outdated-latest.js

```diff
@@ -0,0 +1,111 @@
+'use strict'
+const path = require('path')
+const test = require('tap').test
+const mr = require('npm-registry-mock')
+const Tacks = require('tacks')
+const File = Tacks.File
+const Dir = Tacks.Dir
+const extend = Object.assign || require('util')._extend
+const common = require('../common-tap.js')
+
+const basedir = path.join(__dirname, path.basename(__filename, '.js'))
+const testdir = path.join(basedir, 'testdir')
+const cachedir = path.join(basedir, 'cache')
+const globaldir = path.join(basedir, 'global')
+const tmpdir = path.join(basedir, 'tmp')
+
+const conf = {
+  cwd: testdir,
+  env: extend(extend({}, process.env), {
+    npm_config_cache: cachedir,
+    npm_config_tmp: tmpdir,
+    npm_config_prefix: globaldir,
+    npm_config_registry: common.registry,
+    npm_config_loglevel: 'warn'
+  })
+}
+
+function exampleManifest (version) {
+  return {
+    name: 'example',
+    version: version
+  }
+}
+
+const examplePackument = {
+  'name': 'example',
+  'dist-tags': {
+    'latest': '1.2.4',
+    'beta': '1.2.6'
+  },
+  'versions': {
+    '1.2.0': exampleManifest('1.2.0'),
+    '1.2.1': exampleManifest('1.2.1'),
+    '1.2.2': exampleManifest('1.2.2'),
+    '1.2.3': exampleManifest('1.2.3'),
+    '1.2.4': exampleManifest('1.2.4'),
+    '1.2.5': exampleManifest('1.2.5'),
+    '1.2.6': exampleManifest('1.2.6')
+  }
+}
+
+const fixture = new Tacks(Dir({
+  cache: Dir(),
+  global: Dir(),
+  tmp: Dir(),
+  testdir: Dir({
+    node_modules: Dir({
+      example: Dir({
+        'package.json': File({
+          name: 'example',
+          version: '1.2.3'
+        })
+      })
+    }),
+    'package.json': File({
+      name: 'outdated-latest',
+      version: '1.0.0',
+      dependencies: {
+        example: '^1.2.0'
+      }
+    })
+  })
+}))
+
+function setup () {
+  cleanup()
+  fixture.create(basedir)
+}
+
+function cleanup () {
+  fixture.remove(basedir)
+}
+
+let server
+
+test('setup', function (t) {
+  setup()
+  mr({port: common.port, throwOnUnmatched: true}, function (err, s) {
+    if (err) throw err
+    server = s
+    server.get('/example').reply(200, examplePackument)
+    t.done()
+  })
+})
+
+test('example', function (t) {
```

Comment:
what would be the expected behavior when a user runs `npm upgrade` if `beta` stayed at `1.2.6`, installed/wanted was `^1.2.1` and then `2.0.0` had been released as `latest`?

Reply:
This is the algorithm npm uses to pick versions: https://github.com/zkat/npm-pick-manifest/blob/latest/index.js

The "tag" is a special value that happens to be a cutoff if the algorithm happens to run into it.

My expectation in this case is that `1.2.6` is what would be installed if your `package.json` says `^1.2.1` for that dependency, since `latest` wouldn't be among those. But if you do `npm i pkg@latest`, then you'd get `2.0.0`.

Reply:
thanks - not trying to be a pain, but think this means that a user can still get sort of accidentally steered in to a prerelease/alternative dist-tag using `npm upgrade` unless the package maintainer takes some kind of preventative measure*

it boils down to semver being assumed to be linear, but dist-tags are trying to create branches, I think?

* (example that is obviously too complicated to actually adopt: always publish a semver-patch non prerelease version after pre-releases -- so, if npm wanted to release 6 as the next thing after 5.7.1, they should actually publish 5.7.2 from their non-prelease branch to add a guard for users at ^5 that are doing upgrade)

^ this is silly, and the docs do say `npm i -g npm@latest` is the right upgrade path, I'm just not sure how on would document `update` in such a way that people will know what to expect

Reply:
The only tag that has meaning to npm is `latest` so all logic must be expressed in terms of that. All other tags are arbitrary.

Reply:
I see. Would a `--safe` flag that only updates if `latest` semver matches wanted be useful? Should it be the default? This would bail in the above scenario rather than installing a pre-release/beta/rc version on a users system that had only ever done `npm i foo@latest` and then (some time later) `npm upgrade foo`

Maybe I am obsessing over a corner case here, and the takeaway should be to only install `latest` or a specifically known version that has been released.

Reply:
The entire point of this patch is to make it so that the user who does `npm upgrade foo` doesn't get a pre-release version, so I think we're pretty in alignment here. =)